### DISCUSSION TITLE
Which client is recommended for vanilla JS front-end with subscription operations supported.

### DISCUSSION THREAD

**Discussion by UchihaYuki** [dotansimha/graphql-code-generator#9984] [Q&A] [answered]
I used `graphql-request`, but I found it doesn't support subscription operations. Maybe they do:
https://github.com/jasonkuhrt/graphql-request/pull/330/files
But they lack documentation. 

Please recommend one with rich documentation.

**Comment by n1ru4l** [comment, marked as answer]
The solution depends on your protocol.

For GraphQL over WS we recommend [graphql-ws](https://the-guild.dev/graphql/ws/get-started#use-the-client).

For GraphQL over SSE we recommend [graphql-sse](https://the-guild.dev/graphql/sse/get-started#use-the-client).

**Reply by UchihaYuki** [reply]
It seems [graphql-ws](https://the-guild.dev/graphql/ws/get-started#use-the-client) is what I want.

Are there any plugins from `graphql-code-generator` that can be used with `graphql-ws` to make the development flow easier?

It seems I can use `typescrpt` and `typescript-operations` plugins to have typings in my client code. I'm still wondering if there is a plugin like `graphql-request` for `graphql-request` library to make it even more simpler.

**Reply by n1ru4l** [reply]
@UchihaYuki I recommend `client` preset `documentMode: 'string'`.

https://the-guild.dev/graphql/codegen/plugins/presets/preset-client#documentmode

You can setup a custom fetcher that wraps `graphql-ws` based on this guide: https://the-guild.dev/graphql/codegen/docs/guides/react-vue#appendix-i-react-query-with-a-custom-fetcher-setup
